vkrunner/vr-pipeline.c

```diff
@@ -709,30 +730,94 @@ create_vk_descriptor_set_layout(struct vr_pipeline *pipeline,
+        ++n_desc_sets;
+
+        VkDescriptorPoolSize pool_sizes[2] = {0};
+        uint32_t n_pool_sizes = 0;
+        if (n_ubo) {
+                pool_sizes[n_pool_sizes].type =
+                        VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER;
+                pool_sizes[n_pool_sizes].descriptorCount = n_ubo;
+                n_pool_sizes++;
+        }
+        if (n_ssbo) {
+                pool_sizes[n_pool_sizes].type =
+                        VK_DESCRIPTOR_TYPE_STORAGE_BUFFER;
+                pool_sizes[n_pool_sizes].descriptorCount = n_ssbo;
+                n_pool_sizes++;
+        }
+
+        VkDescriptorPoolCreateInfo descriptor_pool_create_info = {
+                .sType = VK_STRUCTURE_TYPE_DESCRIPTOR_POOL_CREATE_INFO,
+                .flags = VK_DESCRIPTOR_POOL_CREATE_FREE_DESCRIPTOR_SET_BIT,
+                .maxSets = n_desc_sets,
+                .poolSizeCount = n_pool_sizes,
+                .pPoolSizes = pool_sizes
         };
 
-        VkDescriptorSetLayout descriptor_set_layout;
-
-        res = vkfn->vkCreateDescriptorSetLayout(pipeline->window->device,
-                                                &create_info,
-                                                NULL, /* allocator */
-                                                &descriptor_set_layout);
-
-        vr_free(bindings);
+        pipeline->descriptor_set_layout = NULL;
 
+        res = vkfn->vkCreateDescriptorPool(pipeline->window->device,
+                                           &descriptor_pool_create_info,
+                                           NULL, /* allocator */
+                                           &pipeline->window->context->
+                                           descriptor_pool);
         if (res != VK_SUCCESS) {
                 vr_error_message(pipeline->window->config,
-                                 "Error creating descriptor set layout");
-                return VK_NULL_HANDLE;
+                                 "Error creating VkDescriptorPool");
+                goto error;
         }
 
-        return descriptor_set_layout;
+        VkDescriptorSetLayout *descriptor_set_layout =
+                vr_alloc(sizeof(VkDescriptorSetLayout) * n_desc_sets);
+        unsigned *first_sets = vr_alloc(sizeof(unsigned) * n_desc_sets);
+
+        for (unsigned i = 0; i < n_desc_sets; i++) {
+                VkDescriptorSetLayoutCreateInfo create_info = {
+                        .sType =
+                         VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO,
+                        .bindingCount = info[i].n_bindings,
+                        .pBindings = info[i].bindings
+                };
+
+                res = vkfn->vkCreateDescriptorSetLayout(
+                                pipeline->window->device,
+                                &create_info,
+                                NULL, /* allocator */
+                                &descriptor_set_layout[i]);
+
```

Comment:
If this fails the descriptor set layouts created up to this point along with the descriptor_set_layout array will be leaked. Maybe it would be better to directly store descriptor_set_layout on the vr_pipeline when it is allocated so that vr_pipeline_free can clean it up on error. It would have to be allocated with vr_calloc and vr_pipeline_free would have to check for VK_NULL_HANDLE when freeing them.

Reply:
I did not consider this exception. Thanks.
I  changed `VkDescriptorSetLayout *descriptor_set_layout = vr_alloc(sizeof(VkDescriptorSetLayout) * n_desc_sets);` to `pipeline->descriptor_set_layout = vr_calloc(...);` and added check for VK_NULL_HANDLE when freeing them.